### PR TITLE
Update to Node 18.0.0

### PR DESCRIPTION
**Pull request by ShogunPanda**
Summary:

1. Rewrite the build system reducing dependency.
2. Drop compatiblity for Node < 12.0 and Internet Explorer. Only active Node versions and ever-green browsers are supported.
3. Dropped most dependency (note: once we decide to drop support for Node 12 and Node 14, no dependency besides `object-inspect`, which is for browsers, will be needed)
4. Updated the exported `stream` API to Node 18.0.0.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -9,14 +9,14 @@ jobs:
...
-      uses: actions/setup-node@v1
-      with:
-        node-version: ${{ matrix.node-version }}
-    - name: npm install
-      run: npm install
-    - name: npm run test
-      run: npm run test
+      - uses: actions/checkout@v1
+      - name: Use Node.js ${{ matrix.node-version }} on ${{ matrix.os }}
+        uses: actions/setup-node@v1
```
Comment (aduh95) [inline .github/workflows/ci.yml:16]:
```suggestion
        uses: actions/setup-node@v3
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -9,14 +9,14 @@ jobs:
...
-    - uses: actions/checkout@v1
-    - name: Use Node.js ${{ matrix.node-version }} on ${{ matrix.os }}
-      uses: actions/setup-node@v1
-      with:
-        node-version: ${{ matrix.node-version }}
-    - name: npm install
-      run: npm install
-    - name: npm run test
-      run: npm run test
+      - uses: actions/checkout@v1
```
Comment (aduh95) [inline .github/workflows/ci.yml:14]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/sauce.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ jobs:
       - name: Set up node
         uses: actions/setup-node@v1
```
Comment (aduh95) [inline .github/workflows/sauce.yml:10]:
```suggestion
        uses: actions/setup-node@v3
```

File: .gitignore [hunk 4]
```diff
@@ -1,8 +1,4 @@
...
 node_modules/
-.zuul.yml
-.nyc_output
-coverage
-package-lock.json
-*.tap
-.airtaprc
-yarn.lock
+node-*.tar.gz
+package-lock.json
```
Comment (aduh95) [inline .gitignore:4]:
```suggestion
package-lock.json

```

File: lib/internal/js_stream_socket.js [hunk 5]
```diff
@@ -0,0 +1,242 @@
+'use strict';
```
Comment (mcollina) [inline lib/internal/js_stream_socket.js:1]: remove this file, it's not needed

File: lib/internal/inspect.js [hunk 6]
```diff
@@ -0,0 +1,2299 @@
+'use strict';
```
Comment (mcollina) [inline lib/internal/inspect.js:1]: remove this file, it's not needed

File: build/build.mjs [hunk 7]
```diff
@@ -0,0 +1,192 @@
+import { createReadStream } from 'node:fs'
+import { copyFile, cp, mkdir, readFile, rm, writeFile } from 'node:fs/promises'
+import { dirname, resolve } from 'node:path'
```
Comment (aduh95) [inline build/build.mjs:3]:
```suggestion
import { dirname, resolve } from 'node:path'
import process from 'node:process';
```

File: lib/internal/inspect-browser.js [hunk 8]
```diff
@@ -0,0 +1,2299 @@
+'use strict';
```
Comment (mcollina) [inline lib/internal/inspect-browser.js:1]: remove this file, it's not needed

File: examples/capslock-type.cjs [hunk 9]
```diff
@@ -0,0 +1,29 @@
+const { Transform } = require('../lib')
```
Comment (aduh95) [inline examples/capslock-type.cjs:1]:
```suggestion
'use strict'

const process = require('node:process')
const { Transform } = require('../lib')
```

File: lib/internal/errors.js [hunk 10]
```diff
@@ -0,0 +1,1663 @@
+/* eslint node-core/documented-errors: "error" */
```
Comment (mcollina) [inline lib/internal/errors.js:1]: Please use the previous `error.js` replacement instead of this from core.

File: examples/typer.mjs [hunk 11]
```diff
@@ -0,0 +1,22 @@
+import { createReadStream } from 'node:fs'
```
Comment (aduh95) [inline examples/typer.mjs:1]:
```suggestion
import { createReadStream } from 'node:fs'
import process from 'node:process'
```

File: lib/internal/primordials.js [hunk 12]
```diff
@@ -0,0 +1,446 @@
+
+    'use strict';
```
Comment (mcollina) [inline lib/internal/primordials.js:2]: remove this file, we don't want primordials.

File: package.json [hunk 13]
```diff
@@ -2,69 +2,73 @@
...
-    "test-browsers-local": "airtap -p local test/browser.js",
-    "cover": "nyc npm test",
-    "report": "nyc report --reporter=lcov",
-    "update-browser-errors": "babel -o errors-browser.js errors.js"
-  },
-  "repository": {
-    "type": "git",
-    "url": "git://github.com/nodejs/readable-stream"
-  },
+  "author": "NearForm Ltd",
```
Comment (mcollina) [inline package.json:5]: That's not correct

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by benjamingr** [comment]
This is great, awesome work - thanks 🙏 

(Not LGTMing since I didn't review fully and Matteo is already looking whom I trust to approve once issues have been resolved)

**Comment by mcollina** [comment]
If supporting Node v12 does not add much of a problem, then it would be great!

**Comment by ShogunPanda** [comment]
I'll try and see what happens. Will keep you posted!

File: .gitignore (hunk 4 above)
Comment (ShogunPanda) [reply .gitignore]: 👍

File: lib/internal/primordials.js (hunk 12 above)
Comment (ShogunPanda) [reply lib/internal/primordials.js]: Unfortunately the Node core uses primordials everywhere.
I've replaced with a custom slimmed down primordials code.

**Comment by mcollina** [comment]
Good work! I see CI is passing.. minus Safari. Do you know why?

**Comment by ShogunPanda** [comment]
There is a bug reported in airtap-sauce (https://github.com/airtap/sauce/issues/11).

I was thinking: since we can use playwright for all browsers and all platforms (thanks to GH actions matrix), do we still need sauce?

Anyway, tomorrow I'll focus on Node 12 compatibility (I think I just have to transpile `?.` and `??`) and this should be good to go.

**Comment by mcollina** [comment]
playwright would be awesome

**Comment by ShogunPanda** [comment]
@mcollina I was able to accomplish both: Node 12 is now supported and SauceLabs has been dropped in favor of playwright.

File: lib/internal/streams/pipeline.js [hunk 14]
```diff
@@ -1,97 +1,416 @@
...
+  codes: { ERR_INVALID_ARG_TYPE, ERR_INVALID_RETURN_VALUE, ERR_MISSING_ARGS, ERR_STREAM_DESTROYED },
+  AbortError
+} = require('../../ours/errors')
+
+const { validateFunction, validateAbortSignal } = require('../validators')
+
+const { isIterable, isReadable, isReadableNodeStream, isNodeStream } = require('./utils')
+
+if (typeof AbortController === 'undefined') {
+  globalThis.AbortController = require('abort-controller').AbortController
```
Comment (mcollina) [inline lib/internal/streams/pipeline.js:26]: We can't add globals as a side effect of loading readable-stream

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/ours/util.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor
+
+if (typeof Blob === 'undefined') {
+  let { Blob } = require('buffer')
+
+  if (typeof Blob === 'undefined') {
+    Blob = require('blob-polyfill').Blob
+  }
+
+  globalThis.Blob = Blob
```
Comment (mcollina) [inline lib/ours/util.js:12]: Do not alter `globalThis`

File: lib/ours/util.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        if (err) {
+          return reject(err)
+        }
+
+        return resolve(...args)
+      })
+    })
+  },
+
+  inspect: require('object-inspect'),
```
Comment (mcollina) [inline lib/ours/util.js:75]: This should be removed.

File: lib/ours/errors.js [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+  This file is a reduced and adapted version of the main lib/internal/errors.js file defined at
+
+  https://github.com/nodejs/node/blob/master/lib/internal/errors.js
+
+  Don't try to replace with the original file and keep it up to date (starting from E(...) definitions)
+  with the upstream file.
+*/
+
+if (typeof AggregateError === 'undefined') {
+  globalThis.AggregateError = require('aggregate-error')
```
Comment (mcollina) [inline lib/ours/errors.js:12]: Do not install AggregateError as a global

File: lib/ours/errors.js [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+const assert = require('assert')
+
+const { inspect, format } = require('util')
```
Comment (mcollina) [inline lib/ours/errors.js:17]: please avoid loading util.

File: lib/ours/errors.js [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+
+  Don't try to replace with the original file and keep it up to date (starting from E(...) definitions)
+  with the upstream file.
+*/
+
+if (typeof AggregateError === 'undefined') {
+  globalThis.AggregateError = require('aggregate-error')
+}
+
+const assert = require('assert')
```
Comment (mcollina) [inline lib/ours/errors.js:15]: please avoid loading assert

**Review by mcollina — CHANGES_REQUESTED** [review]

File: package.json [hunk 20]
```diff
@@ -2,69 +2,71 @@
...
+    "build": "node build/build.mjs",
+    "postbuild": "prettier -w lib test",
+    "test": "tap --rcfile=./tap.yml test/parallel/test-*.js test/ours/test-*.js",
+    "test:browsers": "airtap -p all test/browser/test-*.js",
+    "coverage": "c8 -c ./c8.json tap --rcfile=./tap.yml test/parallel/test-*.js test/ours/test-*.js",
+    "format": "prettier -w src lib test",
+    "lint": "eslint src"
   },
-  "license": "MIT"
+  "dependencies": {
```
Comment (mcollina) [inline package.json:44]: I think this is missing the string_decoder module for Browser usage. Please verify everything is working webpack and rollup.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/internal/streams/pipeline.js (hunk 14 above)
Comment (ShogunPanda) [reply lib/internal/streams/pipeline.js]: 👍

File: lib/ours/util.js (hunk 15 above)
Comment (ShogunPanda) [reply lib/ours/util.js]: 👍

File: lib/ours/errors.js (hunk 17 above)
Comment (ShogunPanda) [reply lib/ours/errors.js]: 👍

File: lib/ours/util.js (hunk 16 above)
Comment (ShogunPanda) [reply lib/ours/util.js]: 👍

File: lib/ours/errors.js (hunk 18 above)
Comment (ShogunPanda) [reply lib/ours/errors.js]: 👍

File: lib/ours/errors.js (hunk 19 above)
Comment (ShogunPanda) [reply lib/ours/errors.js]: 👍

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**